### PR TITLE
61 navigate directories

### LINKED ISSUE #28: Folder QAbtractTableModel can be slow
After implement the number of pts and dimension in the table model, loading is noticeably slower.

```python
        self.actions_library = {
            "Prefix": lambda file: file.rsplit("_", 1)[0],
            "Scan #": lambda file: int(file.rsplit("_", 1)[1].split(".")[0]),
            "Points": lambda file: self.get_file_pts(file),
            "Dim": lambda file: self.get_file_dim(file),
            "Size": lambda file: self.get_file_size(file),
            "Date": lambda file: self.get_file_date(file),
        }

        self.columnLabels = list(self.actions_library.keys())
```

with:

```python
    def get_file_pts(self, file):
        filepath = self.get_file_path(file)
        return readMDA(str(filepath))[1].curr_pt

    def get_file_dim(self, file):
        filepath = self.get_file_path(file)
        return readMDA(str(filepath))[1].dim
```

### LINKED ISSUE #41: Keep selected POS and DET(s) when switching from one file to another

### LINKED ISSUE #61: Impossible to compare data from different directory
At the moment changing directory refresh the app

### LINKED ISSUE #87: Don't auto-plot the first file
@apsIEX think it is a bad idea.

### LINKED ISSUE #88: Refactor MDAFolderTableModel
Very inefficient: 

1. reads the MDA file 5 times to get pts/dim/pos/date/size, instead of reading it once to gather all 5 info.
2. It also seems to have too many get/set method that might not be even used.

### PR DESCRIPTION
**Pull request by rodolakis**
- close #28
- close #41
- close #61 
- close #87
- close #88 

Sorry, very messy PR, I broke everything and had to change a lot of the logic to make it work again. Past me made a lot of bad design choices if you as present me. 

But now the behavior seems to be what I intended:

When switching directory, `MVC` (Model View Controller) is not systematically cleared: 
- the content of the table views (`MDAFolderTableView` and `SelectFieldsTableView`) is updated accordingly:
  - empty table view  is nothing to display (using new `EmptyTableModel` class), instead of clearing everything. 
  - updated content if something to display 
- the content of the `MDAFileVisualization` widget remains unchanged, allowing to compare files from different directories; opening an invalid or empty directory by mistake does not clear the graph, just displays empty tables.
- only the very first mda file is plotted automatically (when the application is started); subsequent change of folder don't trigger automatic plotting of the first file (#87).

Image

### CODE REVIEW

File: mdaviz/mainwindow.py [hunk 1]
```diff
@@ -229,11 +226,21 @@ def get_all_subfolders(
...
 
             else:
-                self._folderPath = None
-                self._dataPath = None
-                self.setSubfolderList([])
                 comment = f"{str(folder_path)!r} - invalid path."
-                self.folderNotValid(layout, comment)
+                self.folderNotValid(comment)
+
+    def folderNotValid(self, comment):
```
Comment (prjemian) [inline mdaviz/mainwindow.py:232]: This name (`folderNotValid`) implies a test that returns a boolean.

Comment (rodolakis) [reply mdaviz/mainwindow.py:232]: does `ifFolderNotValid` work better?

Comment (prjemian) [reply mdaviz/mainwindow.py:232]: Looks it is [called ](https://github.com/BCDA-APS/mdaviz/blob/75c536fb9960840d59a9523c8a898df6aef6b20d/mdaviz/mainwindow.py#L230)once a selected folder has been determined as not valid.

Comment (rodolakis) [reply mdaviz/mainwindow.py]: I just got rid of that method since it is only called once.

File: mdaviz/mda_folder_table_model.py [hunk 2]
```diff
@@ -11,30 +11,17 @@
...
-            "Date": lambda file: self.get_file_date(file),
-            "Size": lambda file: self.get_file_size(file),
-        }
-
-        self.columnLabels = list(self.actions_library.keys())
-
-        self.setAscending(True)
-        self._folderCount = 0
-
+        # parent = <mdaviz.mda_folder.MDA_MVC object at 0x1101e7520>
```
Comment (prjemian) [inline mdaviz/mda_folder_table_model.py:19]: Consider converting this comment into a docstring that describes the `parent` parameter.  Here's an example:

```
    """
    Create the model and connect with its parent.

    PARAMETERS

    parent object:
        Instance of mdaviz.mda_folder.MDAMVC
    """
```

File: mdaviz/select_fields_table_view.py [hunk 3]
```diff
@@ -20,6 +20,8 @@
 from .select_fields_table_model import TableColumn
 from .select_fields_table_model import TableField
 
+HEADERS = "Field", "X", "Y", "I0", "Unscale", "PV", "DESC", "Unit"
```
Comment (prjemian) [inline mdaviz/select_fields_table_view.py:23]: Would `"Monitor"` be a more general term here?  `"I0"` pulls my attention to a specific `I0` signal when counting time, I00, I000, or some other signal might be selected.

File: mdaviz/utils.py [hunk 4]
```diff
@@ -119,6 +119,35 @@ def get_md(mda_file_metadata):
     return new_metadata
 
 
+def mda2ftm(selection):
+    """Goes from  {'Y': [2, 3], 'X': 1} (mda) to {1: 'X', 2: 'Y', 4: 'Y'} (field TM)"""
```
Comment (prjemian) [inline mdaviz/utils.py:123]: I don't understand this comment.  Transforms `selection` into ...(please explain)...

File: mdaviz/utils.py [hunk 5]
```diff
@@ -119,6 +119,35 @@ def get_md(mda_file_metadata):
...
+    if selection is not None:
+        ftm_selection = {
+            v: "X" if k == "X" else "Y"
```
Comment (prjemian) [inline mdaviz/utils.py:126]: Are there only two choice for `k`?  Why not use `k` here?

File: mdaviz/mda_folder_table_model.py (hunk 2 above)
Comment (prjemian) [reply mdaviz/mda_folder_table_model.py:19]: And apply to other similar instances.  Alternative is to rename the parameter from `parent` to `mvamvc_parent`.

Comment (rodolakis) [reply mdaviz/mda_folder_table_model.py:19]: That comment was a note to myself during debugging. I was trying to keep track of the parent-child relationships. I can update the docstrings. I did rename `self.parent` to `self.mda_mvc`.

Comment (rodolakis) [reply mdaviz/mda_folder_table_model.py]: I updated all the classes with docstrings https://github.com/BCDA-APS/mdaviz/pull/91/commits/27b27ba6ad5c436157386b18be78cd1d215e6780

File: mdaviz/select_fields_table_view.py (hunk 3 above)
Comment (rodolakis) [reply mdaviz/select_fields_table_view.py]: Done

File: mdaviz/utils.py (hunk 4 above)
Comment (rodolakis) [reply mdaviz/utils.py:123]: added docstring

File: mdaviz/utils.py [hunk 6]
```diff
@@ -119,6 +119,35 @@ def get_md(mda_file_metadata):
...
+            for k, vals in selection.items()
+            for v in ([vals] if isinstance(vals, int) else vals)
+        }
+    else:
+        ftm_selection = {}
+    return ftm_selection
+
+
+def ftm2mda(selection):
+    """Goes from  {1: 'X', 2: 'Y', 4: 'Y'} (field TM) to {'Y': [2, 3], 'X': 1} (mda)"""
```
[the remark this answers is not in the document]
Comment (rodolakis) [reply mdaviz/utils.py:136]: added docstring

**Review by prjemian — APPROVED** [review]
We reviewed as a demo.  I'm happy with this code as it is now.

**Pull request merged**